Drools 2976 - HOT RELOAD for drools-wb-screens

gitgabrio opened the pull request:
@mdproctor @Rikkola @porcelli @manstis @jsoltes @danielezonca @etirelli 
Scope of this PR is to enable Hot Reload (SDM) for drools-wb screens when drools-wb-webapp is launched by maven.
This feature is needed to
1) remove dependency from proprietary plugin
2) easy development/debug with low-level machines (launching with maven it is possible to make debug/development without IDE)
This is the first step toward a more general solution.
Things to test:

- Launching with maven "mvn clean process-resources gwt:run"
Debug in chrome
Debug with IntelliJ JS debugger
Hot reload of drools-screen client code (java/resources)
Complete clean of "target" directory

- Launching with IntelliJ plugin
Debug in chrome
Debug with IntelliJ JS debugger
Hot reload of drools-screen client code (java/resources)

This PR depends on https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/817

kiereleaseuser3 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser3 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

Rikkola commented:
@mbiarnes @manstis do you or do you know anyone who might be able to add @gitgabrio to the auto run test list. ( made up the name of the list )

Rikkola commented on drools-wb-webapp/pom.xml:
```diff
@@ -35,6 +35,8 @@
     <!-- Add the absolute path for $JBOSS_HOME below to manage another instance -->
     <errai.jboss.home>${project.build.directory}/wildfly-${version.org.jboss.errai.wildfly}</errai.jboss.home>
+    <!-- This must be overridden on a per-profile base, to meet individual' machine setup -->
+    <path.to.kie.wb.common>../../kie-wb-common</path.to.kie.wb.common>
```
Ok. Based on our chat this is not needed.

Rikkola commented on drools-wb-webapp/pom.xml:
```diff
@@ -1246,26 +1261,199 @@
+
+
+                <!--<source>${path.to.kie.wb.common}/kie-wb-common-widgets/kie-wb-common-ui/src/main/java</source>-->
+                <!--<source>${path.to.kie.wb.common}/kie-wb-common-widgets/kie-wb-common-ui/src/main/resources</source>-->
```
Just clean up these commented bits. The files tend to get too long even with regular content.

Rikkola commented on drools-wb-webapp/pom.xml:
```diff
@@ -1246,26 +1261,199 @@
         <configuration>
           <deploy>${project.build.directory}/gwt-symbols-deploy</deploy>
-          <extraJvmArgs>-Xmx5G -Xms1024m -Xss1M -XX:CompileThreshold=7000 -Derrai.jboss.home=${errai.jboss.home} -Derrai.marshalling.server.classOutput=${project.build.outputDirectory} -Derrai.dynamic_validation.enabled=true</extraJvmArgs>
+          <!-- -Djava.io.tmpdir=${project.build.directory} so that gwt cache directory and all other temporary files gets cleaned on mvn clean !-->
```
More commented lines to clean up.

Rikkola commented on drools-wb-webapp/pom.xml:
```diff
@@ -1331,37 +1519,51 @@
             <compileSourcesArtifact>org.drools:drools-wb-enum-editor-api</compileSourcesArtifact>
-            <compileSourcesArtifact>org.drools:drools-wb-enum-editor-client</compileSourcesArtifact>
+            <!-- Not needed - add as sources !-->
+            <!--<compileSourcesArtifact>org.drools:drools-wb-enum-editor-client</compileSourcesArtifact>-->
```
Here too a lot of commented modules. These will get outdated at one point and since they are not tied to the compile they start to be misleading.

gitgabrio commented:
@Rikkola @danielezonca 
Fixed with
[DROOLS-2976] Removing comments from pom. Renamed Readme file. Add note to point to inotify configuration
[DROOLS-2976] Removing comments from pom.

jsoltes commented:
@Rikkola I think the problem is @gitgabrio isn't in our organization. @manstis do you know how to add him there? He already works for Red Hat for several months, so I don't know why we still see "Can one of the admins verify this PR? Comment with 'ok to test' to start the build."

Rikkola commented:
@jsoltes @manstis Edson already added @gitgabrio

Rikkola commented:
Can't really run it. On the other hand I can't run most of the webapps from CLI. So most likely this is a local problem.

Rikkola commented:
Idea works.

danielezonca commented on drools-wb-webapp/pom.xml:
```diff
@@ -1246,26 +1259,192 @@
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>build-helper-maven-plugin</artifactId>
+        <version>3.0.0</version>
```
Please remove the version to use the one specified in the bom

gitgabrio replied:
Fixed with

[DROOLS-2976] Removing build-helper-maven-plugin version

gitgabrio commented:
@jsoltes 👍

manstis commented:
I guess once https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/817 has its merge conflict resolved these PRs are good to be merged :-)

gitgabrio commented:
@manstis could you merge this one after https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/817, please?

gitgabrio commented:
@manstis could you merge this one please? I've already merged with the "ECJ-FIX". Thx!

manstis commented:
@gitgabrio Do you know if anyone has tested this across multiple repositories? (e.g. ```kie-wb-common``` and ```drools-wb```). @ederign was asking before (presumably) he approves (he's the last reviewer).

gitgabrio commented:
@manstis @ederign @porcelli 
I've tested it with kie-wb-common, and of course it  works (GWT just need access to sources in its "classpath"). 
For external modules there are two issues, though:
1) checkstyle fails with other projects, so you should disable it
2) since the full path of external source is referenced, *as it is* we should all have the same sources in the same path.
About issue 1) there is a ticket https://issues.jboss.org/browse/AF-1486, but I am the only one fixing that
About issue 2) my idea is to have this managed with personal settings, so that everyone could choose what "hot reload" inside its own settings.xml; but I need some time to fix that.
For the moment being, this PR is only about Drools-wb development and inner modules.

danielezonca commented:
@ederign 
Hi, do you have other questions or can we merge?
I think it is better to discuss about external module hot reload in another PR

@gitgabrio 
Can you please update the code from master and ask for a new build?

ederign commented:
Hi @danielezonca,

I don't think we should merge this because we are introducing 'noise' to the pom file. My suggestion would be to go for the solution that I've proposed on "An alternative for hot reload without IDEA plugin"  mail. I'll work on this on this sprint.

But if @porcelli and @manstis agree with this PR I'm outvoted.

etirelli commented:
@ederign define 'noise'? if it is under a maven plugin configuration that has no impact on overall general purpose maven builds, but improves the work for other team members, it might be good for approval. Happy to discuss it on IRC/BJ.

porcelli commented on drools-wb-webapp/pom.xml:
```diff
@@ -1262,26 +1275,191 @@
+            <configuration>
+              <sources>
+                <!-- Include screens sources !-->
+                <source>../drools-wb-screens/drools-wb-drl-text-editor/drools-wb-drl-text-editor-client/src/main/java</source>
```
isn't it missing drools-wb-factmodel-editor-api here?

porcelli replied:
k, makes sense

porcelli commented:
Quick recap of this PR:

 - New dependencies were added, basically to please enforcer, as mostly dependencies would be ok via transient resolution
   -  However, not sure if enforcer makes sense for webapp, as they're just agregators
 - Any new module should be added here (not much different than today)
 - And mostly imporant, As @Rikkola pointed, this might have performance implications
   - We could use profile, but not sure if I like the idea, as we'll have a very tedious change spread in 2 places

 With the above comments and awareness, +1 to merge and we revisit it when the new solution from @ederign be ready.

porcelli commented:
@gitgabrio i just meant that using the sources might create performance issues, but I'm ok with that for now. 

(a way to avoid it, would create a new profile, but this is something that I don't like the idea)

@gitgabrio better?

gitgabrio commented:
@porcelli 
Thanks, I still can not completely understand why and where using sources might create performance issues, but probably because I do not know exactly where the drools-webapp is built and used (beside during development): my questions are just to understand what I am missing of the whole picture.

jhrcek commented on drools-wb-webapp/pom.xml:
```diff
@@ -1262,26 +1275,191 @@
       <scope>runtime</scope>
     </dependency>
+
+    <!-- Included dependencies because previously where cross-provided by other components. -->
```
Just to be aware of the proper naming: what you call cross-provided is what maven officially calls "transitive dependencies".

jhrcek commented on drools-wb-webapp/pom.xml:
```diff
@@ -1262,26 +1275,191 @@
+    <dependency>
+      <groupId>org.uberfire</groupId>
+      <artifactId>uberfire-wires-core-client</artifactId>
+      <scope>compile</scope>
```
Compile scope is the default scope. So if you don't declare scope for a dependency, it is automatically compile scoped (just so you know for next time and can save some boilerplate declarations in the future)
